src/nvim/statusline.c

```diff
@@ -894,13 +894,10 @@ int build_statuscol_str(win_T *wp, linenr_T lnum, long relnum, int maxwidth, int
                                fillchar, maxwidth, hlrec, &clickrec, stcp);
   xfree(stc);
 
-  // Allocate and fill click def array if width has changed
-  if (wp->w_status_click_defs_size != (size_t)width) {
```

Comment:
This memory leak actually rarely happens, because there is a typo, and one rarely has such a narrow window. If this is changed to `w_statuscol_click_defs_size` it happens much more frequently (including in existing tests).

Reply:
Only clearing should be necessary I think, this is the diff that seemed correct to me but leaked anyways which is why I left it as is until I figured it out. Nice spot on the typo:
```diff
@@ -887,6 +887,7 @@ int build_statuscol_str(win_T *wp, linenr_T lnum, long relnum, int maxwidth, int
     set_vim_var_nr(VV_LNUM, lnum);
     set_vim_var_nr(VV_RELNUM, relnum);
   }
+  stl_clear_click_defs(wp->w_statuscol_click_defs, wp->w_statuscol_click_defs_size);

   StlClickRecord *clickrec;
   char *stc = xstrdup(wp->w_p_stc);
@@ -895,8 +896,7 @@ int build_statuscol_str(win_T *wp, linenr_T lnum, long relnum, int maxwidth, int
   xfree(stc);

   // Allocate and fill click def array if width has changed
-  if (wp->w_status_click_defs_size != (size_t)width) {
-    stl_clear_click_defs(wp->w_statuscol_click_defs, wp->w_statuscol_click_defs_size);
+  if (wp->w_statuscol_click_defs_size != (size_t)width) {
     wp->w_statuscol_click_defs = stl_alloc_click_defs(wp->w_statuscol_click_defs, width,
                                                       &wp->w_statuscol_click_defs_size);
     stl_fill_click_defs(wp->w_statuscol_click_defs, clickrec, buf, width, false);
```

Refilling the click definitions for each line still feels [redundant](https://github.com/neovim/neovim/issues/21767).

Reply:
> Refilling the click definitions for each line still feels https://github.com/neovim/neovim/issues/21767, unless we actually start storing the click definitions for each row.

Although redundant, now that the condition is fixed it no longer works properly with the above diff(as expected but worked until now due to the typo). Something more similar to the pattern used in `win_redr_custom()` might be better in the meantime, i.e. clear -> alloc -> build -> fill:

```diff
@@ -887,6 +887,9 @@ int build_statuscol_str(win_T *wp, linenr_T lnum, long relnum, int maxwidth, int
     set_vim_var_nr(VV_LNUM, lnum);
     set_vim_var_nr(VV_RELNUM, relnum);
   }
+  stl_clear_click_defs(wp->w_statuscol_click_defs, wp->w_statuscol_click_defs_size);
+  wp->w_statuscol_click_defs = stl_alloc_click_defs(wp->w_statuscol_click_defs, maxwidth,
+                                                    &wp->w_statuscol_click_defs_size);

   StlClickRecord *clickrec;
   char *stc = xstrdup(wp->w_p_stc);
@@ -894,13 +897,7 @@ int build_statuscol_str(win_T *wp, linenr_T lnum, long relnum, int maxwidth, int
                                fillchar, maxwidth, hlrec, &clickrec, stcp);
   xfree(stc);

-  // Allocate and fill click def array if width has changed
-  if (wp->w_status_click_defs_size != (size_t)width) {
-    stl_clear_click_defs(wp->w_statuscol_click_defs, wp->w_statuscol_click_defs_size);
-    wp->w_statuscol_click_defs = stl_alloc_click_defs(wp->w_statuscol_click_defs, width,
-                                                      &wp->w_statuscol_click_defs_size);
-    stl_fill_click_defs(wp->w_statuscol_click_defs, clickrec, buf, width, false);
-  }
+  stl_fill_click_defs(wp->w_statuscol_click_defs, clickrec, buf, width, false);

   return width;
```

Reply:
If the width is equal, `clickrec` is not assigned to anything, then won't the function names be leaked?

Reply:
> clear -> alloc -> build -> fill

This may crash when `maxwidth` is zero while `w_statuscol_click_defs` is not `NULL`. Strange example:
```vim
call setline(1, ['aaa', 'bbb', 'ccc', 'ddd'])
setlocal statuscolumn=a
redraw
setlocal statuscolumn=%!v:lnum==line('$')?'':'%@MyClickFunc@abcd%T'
redraw
```

Reply:
Yeah I noticed, perhaps just go with what you have here.